control-plane-operator/controllers/metrics/metricscontroller.go

```diff
@@ -0,0 +1,208 @@
+package metrics
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/transport"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	promapi "github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+)
+
+var (
+	ComponentLabel = "app"
+
+	ComponentMemoryRSS = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "hypershift_component_memory_rss",
+		Help: "Component memory usage in bytes grouped by control plane component.",
+	}, []string{ComponentLabel})
+)
+
+func init() {
+	metrics.Registry.MustRegister(ComponentMemoryRSS)
+}
+
+// MetricsController is responsible for reading metrics from the management cluster
+// about the hosted control plane and transforming and re-exposing them as more
+// targeted metrics suitable for high level SLO analysis (e.g. via CI runs).
+//
+// To do the work, the controller periodically queries the user workload monitoring
+// route and correlates the metrics with pods.
+type MetricsController struct {
+	podCache             cache.Cache
+	namespace            string
+	userWorkloadEndpoint string
+	log                  logr.Logger
+}
+
+// Options configure the metrics controller.
+type Options struct {
+	// Namespace is the namespace of the hosted control plane.
+	Namespace string
+
+	// PodCache should provide access to pods in Namespace.
+	PodCache cache.Cache
+
+	// UserWorkloadEndpoint is the HTTPS URL of the OCP user workload monitoring
+	// Thanos query endpoint.
+	UserWorkloadEndpoint string
+
+	// Log is a logger.
+	Log logr.Logger
+}
+
+func NewController(opts *Options) (*MetricsController, error) {
+	opts, err := defaultOpts(opts)
+	if err != nil {
+		return nil, err
+	}
+	controller := &MetricsController{
+		podCache:             opts.PodCache,
+		namespace:            opts.Namespace,
+		userWorkloadEndpoint: opts.UserWorkloadEndpoint,
+		log:                  opts.Log,
+	}
+	return controller, nil
+}
+
+func defaultOpts(opts *Options) (*Options, error) {
+	if len(opts.UserWorkloadEndpoint) == 0 {
+		opts.UserWorkloadEndpoint = "https://thanos-querier.openshift-monitoring.svc:9091"
+	}
+	if len(opts.Namespace) == 0 {
+		return nil, fmt.Errorf("namespace is required")
+	}
+	if opts.PodCache == nil {
+		return nil, fmt.Errorf("pod cache is required")
+	}
+	if opts.Log == nil {
+		return nil, fmt.Errorf("log is required")
+	}
+	return opts, nil
+}
+
+// Run starts collecting metrics on interval until ctx is closed. It blocks.
+func (c *MetricsController) Run(ctx context.Context, interval time.Duration) error {
```

Comment:
Why is the `interval` configured here and not in `NewController`? Removing that as arg would also make this a `Runnable` and allow the `Manager` to run it